crypto/dh/dh_key.c

```diff
@@ -56,11 +56,23 @@ static DH_METHOD dh_ossl = {
     NULL
 };
 
+static const DH_METHOD *default_DH_method = &dh_ossl;
+
 const DH_METHOD *DH_OpenSSL(void)
 {
     return &dh_ossl;
 }
 
+void DH_set_default_method(const DH_METHOD *meth)
```

Comment:
Adding a lock on the settor is not sufficient.  You have to lock on the gettor, copy to a local variable, and then release the lock.  Way too much work.  Documenting our thread safety is a HUGE issue.  Not for this PR.